Create HelloWorld

EduStormCrow opened the pull request:
Starting up

mzorz commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+.DS_Store
+/build
+/captures
+.externalNativeBuild
```
`.gitignore` is our friend 👍 
As you work on the project, you may add some other libraries that require some folder not to be added to the repo - here's a nice source were you can find some other `gitignore` files to take inspiration from https://github.com/github/gitignore

EduStormCrow replied:
thanks for the reference, I have added the default one and then will be keeping up to date based on future libraries usage. Addressed in "9c88e53"

mzorz replied:
for the sake of completeness, can you tell me here which's  the commit  hash that addresses it?

mzorz replied:
ah - found it https://github.com/EduStormCrow/Notes/pull/1/commits/9c88e53ec35ecb19e1858d2b85f8526b8a2b3844

mzorz commented on app/build.gradle:
```diff
@@ -0,0 +1,46 @@
+    defaultConfig {
+        applicationId "com.darkwoods.notes"
+        minSdkVersion 21
+        targetSdkVersion 27
```
`targetSdkVersion` should be the latest for a start over project, unless there's a reason to target a different one. Here's some interesting reading about it https://medium.com/google-developers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd and of course the official documentation is always a must read https://developer.android.com/guide/topics/manifest/uses-sdk-element

EduStormCrow replied:
Thanks for the additional info! Changing it to 28 (commit #102f366). But an issue arise by using 28, I had to change 'com.android.support:appcompat-v7:28.0.0-alpha3' (which I believe is the latest one)... only on that library I got the error. Am I missing something? 
Is there an easy place where you can find which libraries go well with which ones? Searched the android documentation "https://developer.android.com/topic/libraries/support-library/revisions" but alpha 3 is not even there, not sure how updated the android documentation is.
Thanks!

mzorz commented on app/build.gradle:
```diff
@@ -0,0 +1,46 @@
+
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
```
kotlin! nice! Let's see what I find out as I keep scrolling down your PR 💯

mzorz replied:
ok I finished scrolling and there were no Kotlin files in here. We should probably remove Kotlin support until we introduce it. You can leave it for the sake of the exercise, the takeaway is that you should only add the dependencies you need. Also, it is OK to say that you plan to use this in the very near future, as an answer to this PR review comment.

EduStormCrow replied:
My idea was to use it on start, but got deprioritized. So I will be removing this from now... (commit #102f366)

mzorz commented on app/build.gradle:
```diff
@@ -0,0 +1,46 @@
+    androidTestImplementation 'com.android.support.test:runner:1.0.2'
+    androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
+    // CardView
+    compile 'com.android.support:cardview-v7:23.3.+'
```
When opening the project in Android Studio, I get a warning about this: 
```
Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html
```

EduStormCrow replied:
Slipped when Copy+Pasting from android documentation... already fixed this on (commit #102f366)

mzorz commented on app/src/androidTest/java/com/darkwoods/notes/ExampleInstrumentedTest.kt:
```diff
@@ -0,0 +1,24 @@
+//        val appContext = InstrumentationRegistry.getTargetContext()
+//        assertEquals("com.darkwoods.notes", appContext.packageName)
+//    }
+}
```
Generally, you shouldn't push commented code to the repo's `master` branch. It's ok to have a feature branch you're woking on with commented code (if the group working there accepts so). As a single developer project, I think it's perfectly fine to create your own rules (you're the owner of the project after all), but just for the sake of getting used to how to work with others this should be taken into account.
As an example these 3 files when you have time, in the WordPress repo:
https://github.com/wordpress-mobile/WordPress-Android/
<img width="247" alt="screen shot 2018-07-19 at 08 14 29" src="https://user-images.githubusercontent.com/6597771/42939377-ce2f04f8-8b2b-11e8-91ab-eb6be3e02629.png">

EduStormCrow replied:
Seems good... will keep that in mind.

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+        DividerItemDecoration decoration = new DividerItemDecoration(getApplicationContext(), VERTICAL);
+        mRecyclerView.addItemDecoration(decoration);
+
+        //TODO agregar un fondo cuando no hay elementos
```
I might sound annoying but let's get used to English please

mzorz replied:
also there's a general unwritten rule to use a space right after the `//`, just the same way you have in the `// TODO` comments below

mzorz commented on app/src/main/java/com/darkwoods/notes/database/NoteDao.java:
```diff
@@ -0,0 +1,34 @@
+    @Query("SELECT * FROM Note ORDER BY creation_date desc")
+    LiveData<List<Note>> loadAllNotes();
+
+       @Insert
```
nitpick: there's an extra `tab` character here that somehow affects readability by drawing attention in an undesirable manner. Can we remove that extra `tab` character please?

mzorz commented on app/src/main/java/com/darkwoods/notes/NotesAdapter.java:
```diff
@@ -0,0 +1,188 @@
+        @Override
+        public boolean onLongClick(View view) {
+            Note mNote = mNotesList.get(getAdapterPosition());
+//            int elementId = mNotesList.get(getAdapterPosition()).getId();
```
this comment should be removed before merging

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+
+        //Note mNote = mAdapter.getNoteOnPosition(itemId);
+        deleteNote(mNote);
+        }
```
extra tab here as well - let's remove it

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+
+
+
+
```
there are 4 newlines here, when 1 should be enough

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+
+        mRecyclerView.setAdapter(mAdapter);
+
+
```
2 newlines here as well, let's remove the extra one

mzorz commented on app/src/main/res/layout/activity_list_notes.xml:
```diff
@@ -0,0 +1,26 @@
+
+    <!-- Floating action button that launches AddTaskActivity when clicked -->
+    <android.support.design.widget.FloatingActionButton
+        android:id="@+id/fab"
```
can we rename the `FloatingActionButton` id from `fab` to something a bit more descriptive? It makes sense to keep the naming convention / structure consistent, for example you have `android:id="@+id/recyclerViewNotes"`  up there. Let's maybe name this `addNoteFAB` or alike?

EduStormCrow replied:
Done (commit #5a21eca)

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+        return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, r.getDisplayMetrics()));
+    }
+}
+
```
let's remove these few extra lines here at EOF

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+        }
+
+    /**
+     * onDeleteNote is called when we want double check the delete of a note form the database
```
did you mean `deleteNote` here or `onDeleteNote`? Also, where does `onDeleteNote` exist?

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+    /**
+     * onDeleteNote is called when we want double check the delete of a note form the database
+     */
+    public void deleteNote(final Note mNote) {
```
we could make this method `private` as it's [only used in `onLongClickListener`](https://github.com/EduStormCrow/Notes/pull/1/files#diff-a7f39c4cd0d6def88c7b6c7605dd2606R156)

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+    @Override
+    public void onLongClickListener(Note mNote) {
+
+        //Note mNote = mAdapter.getNoteOnPosition(itemId);
```
let's remove this commented code

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+
+
+    @Override
+    public void onLongClickListener(Note mNote) {
```
let's not name a method parameter with the `m` prefix, which should be reserved for class `m`embers. A good replacement is just `onLongClickListener(Note note)`.

mzorz replied:
pro-tip: you can select `mNote` and then right click and choose `Refactor->Rename`.  It will change the name everywhere in the code where this variable name is referenced (I know it's overkill for this particular 1-line method, but it'll save you time if you get used to it).

![edurefactor](https://user-images.githubusercontent.com/6597771/42940033-60976f18-8b2e-11e8-95f0-4c01d2c98754.gif)

EduStormCrow replied:
Fixed. Thanks for the Rename tip, I have been using it for a while now... good way of not messing things up.

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+     */
+    public void deleteNote(final Note mNote) {
+
+        final AlertDialog.Builder b = new AlertDialog.Builder(this);
```
single-character variable names are easy to get confused with. Always use some more descriptive names, like `final AlertDialog.Builder builder`.

mzorz commented on app/src/main/java/com/darkwoods/notes/ListNotesActivity.java:
```diff
@@ -0,0 +1,238 @@
+    /**
+     * RecyclerView item decoration - give equal margin around grid item
+     */
+    public class GridSpacingItemDecoration extends RecyclerView.ItemDecoration {
```
if you got this code from somewhere, let's please add a link to the source in the comments

EduStormCrow replied:
I need to find where I took it from, some comment in stackoverflow.com

mzorz commented on app/src/main/java/com/darkwoods/notes/NotesAdapter.java:
```diff
@@ -0,0 +1,188 @@
+    public Note getNoteOnPosition(int pos){
+        return mNotesList.get(pos);
+    }
+
```
extra new lines here

EduStormCrow replied:
fixed in #8630c2

mzorz commented on app/src/main/java/com/darkwoods/notes/NotesAdapter.java:
```diff
@@ -0,0 +1,188 @@
+
+        void onLongClickListener(Note mNote);
+    }
+
```
more extra new lines

EduStormCrow replied:
fixed in #8630c2

mzorz commented on app/src/main/java/com/darkwoods/notes/NotesAdapter.java:
```diff
@@ -0,0 +1,188 @@
+
+
+    }
+
```
many extra new lines

mzorz commented on app/src/main/java/com/darkwoods/notes/database/Note.java:
```diff
@@ -0,0 +1,65 @@
+
+    @PrimaryKey(autoGenerate = true)
+    int id;
+    private String title;
```
we were using the m-prefixed, camel case naming convention elsewhere, why did we change it here? This should be `mTitle` for example. Unless there's a reason for this (there might be one!). Same applies to the other class member variables

EduStormCrow replied:
As those names are used for generating the database, I prefered names without the m-prefix. (I could have also used the m and then add a @ColumnInfo tag, but would have to do it for every field. I thought it was more confusing than what it added in value, also as this is a POJO.

mzorz replied:
Good choice! that explains it. Thanks for clarifying.

mzorz commented on app/src/main/java/com/darkwoods/notes/database/NoteVModel.java:
```diff
@@ -0,0 +1,23 @@
+public class NoteVModel extends ViewModel {
+
+    private LiveData<Note> noteLiveData;
+
```
extra lines

EduStormCrow replied:
fixed #8630c2

mzorz commented on app/src/main/java/com/darkwoods/notes/database/NoteVModel.java:
```diff
@@ -0,0 +1,23 @@
+    private LiveData<Note> noteLiveData;
+
+
+        public NoteVModel(AppDatabase database, int noteID) {
```
extra tabs in this whole constructor method

EduStormCrow replied:
fixed #cd4db64

mzorz commented:
Left a few comments in the code review, also some recommendations for a start:

1 - the PR should have a description, even if a super short one, that describes what it is about.

2 - I'd expect a `README.md` file to exist in the root folder, and a brief explanation of what the project is about. Please add it to your working branch (`start-branch`) and update the remote repo.
There's actually a README.md file in `master`. Your branch should have been branched off from `master` as well. It's ok as this is a first setup.

3 - the notes list is empty at first run, it'd be nice if it showed an empty placeholder or prompt the user to start writing notes, like "Hey! Make your first note bu clicking the + button".

4 - when adding a note, the keyboard goes up and hides the "save" FAB. It took me a fraction of a second to realize, but that might be frustrating for users. You might want to try and think how the FAB can be moved up when the keyboard is up if possible at all, or make the keyboard's default action for the second (last) EditText to be "SAVE" instead of "ENTER".
![edu2](https://user-images.githubusercontent.com/6597771/42941694-edb73694-8b33-11e8-9ec2-14faa664ce93.gif)


Other than that, this looks pretty cool Edu! 😄 I'm excited about the work you've done. Keep it on! 💪

EduStormCrow commented:
Thanks @mzorz  for doing this... I fixed part of the code and added a few comments to your reviews.

Based on your questions...
1) Added the description.
2) I added a readme in the master to had something to compere while doing the merge. I will update that one with a basic description.
3) Yes, that was the first item of the TODO list. I will add that functionality later on.
4) Yes, that was the 3rd item of the TODO list. I will work on that, I believe moving the button will be better as you are adding a description that could have multiple lines, so the enter approach will not be intuitive.
Again, thanks for taking a look at my code!

mzorz commented:
Excellent! Thanks for fixing and taking the exposed considerations into account. 
LGTM :shipit:
